### PR TITLE
fix: Make multi cache store return error upon getting invalid height

### PR DESCRIPTION
**Pull request by mattverse**
Closes:[ #2710](https://github.com/osmosis-labs/osmosis/issues/2710)

## What is the purpose of the change

Querying an invalid height on the multi cache store would not return an error before, instead returning an empty store. 
This causes confusion in queries and many other cases, where we would not know whether the query results are coming from the actual results or the result coming from trying to query an invalid height in the store.

This fix changes it so that trying to get a multi cache store at an invalid height would return error.

This PR also includes fixes for other tests related to stores and snapshots, as the previous tests were not sufficient for testing(as they did not return an error while the tests were error checking)

## Brief Changelog
- Make IAVL store return error upon getting invalid height.

## Testing and Verifying

This PR fixes existing tests
## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: pruning/manager.go [hunk 1]
```diff
@@ -135,7 +135,7 @@ func (m *Manager) HandleHeight(previousHeight int64) int64 {
 
 	if int64(m.opts.KeepRecent) < previousHeight {
 		pruneHeight := previousHeight - int64(m.opts.KeepRecent)
-		// We consider this height to be pruned iff:
```
Comment (ValarDragon) [inline pruning/manager.go]: This is intentional I think

(`iff` is common short hand for `if and only if` in math / logical proof contexts)

File: store/iavl/store.go [hunk 2]
```diff
@@ -106,7 +106,7 @@ func UnsafeNewStore(tree *iavl.MutableTree) *Store {
 // Any mutable operations executed will result in a panic.
 func (st *Store) GetImmutable(version int64) (*Store, error) {
 	if !st.VersionExists(version) {
-		return &Store{tree: &immutableTree{&iavl.ImmutableTree{}}}, nil
+		return nil, fmt.Errorf("version mismatch on immutable IAVL tree; version does not exist")
```
Comment (ValarDragon) [inline store/iavl/store.go]: Lmao, this seems 100% correct, can't believe we did the other thing before lol

Comment (ValarDragon) [inline store/iavl/store.go]:
```suggestion
		return nil, fmt.Errorf("version mismatch on immutable IAVL tree; version does not exist. Version has either been pruned, or is for a future block height.")
```

File: store/rootmulti/store_test.go [hunk 3]
```diff
@@ -504,9 +504,9 @@ func TestMultiStore_Pruning(t *testing.T) {
 		saved       []int64
 	}{
 		{"prune nothing", 10, pruningtypes.NewPruningOptions(pruningtypes.PruningNothing), nil, []int64{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}},
-		{"prune everything", 10, pruningtypes.NewPruningOptions(pruningtypes.PruningEverything), []int64{1, 2, 3, 4, 5, 6, 7, 8, 9}, []int64{10}},
-		{"prune some; no batch", 10, pruningtypes.NewCustomPruningOptions(2, 1), []int64{1, 2, 4, 5, 7}, []int64{3, 6, 8, 9, 10}},
-		{"prune some; small batch", 10, pruningtypes.NewCustomPruningOptions(2, 3), []int64{1, 2, 4, 5}, []int64{3, 6, 7, 8, 9, 10}},
+		{"prune everything", 22, pruningtypes.NewPruningOptions(pruningtypes.PruningEverything), []int64{1, 2, 3, 4, 5, 6, 7, 8, 9}, []int64{10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22}},
+		{"prune some; no batch", 10, pruningtypes.NewCustomPruningOptions(2, 1), []int64{1, 2, 3, 4, 5, 6, 7}, []int64{8, 9, 10}},
+		{"prune some; small batch", 10, pruningtypes.NewCustomPruningOptions(2, 3), []int64{1, 2, 3, 4, 5, 6}, []int64{7, 8, 9, 10}},
```
Comment (ValarDragon) [inline store/rootmulti/store_test.go:509]: kinda confused about why these changes (do they change something for this PR)

File: store/rootmulti/snapshot_test.go [hunk 4]
```diff
@@ -128,7 +128,7 @@ func TestMultistoreSnapshot_Checksum(t *testing.T) {
 			"71a0b594863b63f6341092ceee1a8d508722af4bed2db1fb3f1c58775b95459c",
 			"0864b7d52db6d2e6fee11db2399d042b52179221012eb20f969e28843a2e20ad",
 			"7e57a64b4a344e58fa95549f07a784a1247dca38a7db5b034ee2bd2f2011936f",
-			"534b87533ca86f404b59c34f0dc451b36c02a2dacfc51abf2863668bd40e0d04",
+			"396883880eb125b55f0b880e0bcdb59b4b5ea6d07631e1e8fef8353df748e2b8",
```
Comment (ValarDragon) [inline store/rootmulti/snapshot_test.go:131]: I don't understand why snapshots changed O_O

File: baseapp/baseapp_test.go [hunk 5]
```diff
@@ -413,9 +413,9 @@ func TestLoadVersionPruning(t *testing.T) {
 		lastCommitID = sdk.CommitID{Version: i, Hash: res.Data}
 	}
 
-	for _, v := range []int64{1, 2, 3, 4} {
+	for _, v := range []int64{1, 2, 4} {
 		_, err = app.cms.CacheMultiStoreWithVersion(v)
-		require.NoError(t, err)
```
Comment (ValarDragon) [inline baseapp/baseapp_test.go:418]: whats going on here?

File: baseapp/baseapp_test.go [hunk 6]
```diff
@@ -413,9 +413,9 @@ func TestLoadVersionPruning(t *testing.T) {
 		lastCommitID = sdk.CommitID{Version: i, Hash: res.Data}
 	}
```
Comment (ValarDragon) [inline baseapp/baseapp_test.go]:
```suggestion
// Ensure that load version returns an error for every pruned version
```

File: pruning/manager.go (hunk 1 above)
Comment (alexanderbez) [reply pruning/manager.go]: correct :P

**Review by alexanderbez — APPROVED** [review]

File: store/rootmulti/store_test.go [hunk 7]
```diff
@@ -504,9 +504,9 @@ func TestMultiStore_Pruning(t *testing.T) {
 		saved       []int64
 	}{
 		{"prune nothing", 10, pruningtypes.NewPruningOptions(pruningtypes.PruningNothing), nil, []int64{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}},
-		{"prune everything", 10, pruningtypes.NewPruningOptions(pruningtypes.PruningEverything), []int64{1, 2, 3, 4, 5, 6, 7, 8, 9}, []int64{10}},
-		{"prune some; no batch", 10, pruningtypes.NewCustomPruningOptions(2, 1), []int64{1, 2, 4, 5, 7}, []int64{3, 6, 8, 9, 10}},
-		{"prune some; small batch", 10, pruningtypes.NewCustomPruningOptions(2, 3), []int64{1, 2, 4, 5}, []int64{3, 6, 7, 8, 9, 10}},
+		{"prune everything", 22, pruningtypes.NewPruningOptions(pruningtypes.PruningEverything), []int64{1, 2, 3, 4, 5, 6, 7, 8, 9}, []int64{10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22}},
```
Comment (mattverse) [inline store/rootmulti/store_test.go:507]: this I changed because the test cases that were existing previously were not sufficient to prove that pruning works correctly. Prune everything has keep recent 10 and prune everything interval 10, so we need at least height above 22

File: store/rootmulti/store_test.go [hunk 8]
```diff
@@ -529,7 +529,7 @@ func TestMultiStore_Pruning(t *testing.T) {
 
 			for _, v := range tc.deleted {
 				_, err := ms.CacheMultiStoreWithVersion(v)
-				require.NoError(t, err, "expected error when loading height: %d", v)
+				require.Error(t, err, "expected error when loading height: %d", v)
```
Comment (mattverse) [inline store/rootmulti/store_test.go:532]: this should change because getting deleted height should be getting errors

File: store/rootmulti/store_test.go (hunk 3 above)
Comment (mattverse) [reply store/rootmulti/store_test.go:509]: these change because the second paramter in the NewCustomPruningOptions is `interval` which indicates the interval to check if state should be pruned, not the height that should be pruned. The previous test cases were incorrect

File: store/rootmulti/snapshot_test.go (hunk 4 above)
Comment (mattverse) [reply store/rootmulti/snapshot_test.go:131]: cc @p0mvn do you have an idea why these might have changed from these changes?

Comment (tac0turtle) [reply store/rootmulti/snapshot_test.go:131]: I think since the version that wasn't preset was filled with {} it was accounted for in the checksum, but now since its an error instead of an empty value the hash may change. While its may not be an actual issue on a network I think the test accounts for it

Comment (tac0turtle) [reply store/rootmulti/snapshot_test.go:131]: oh its actually a go 1.19 issue.

Comment (mattverse) [reply store/rootmulti/snapshot_test.go:131]: @marbar3778 wdym by its an go 1.19 issue?

Comment (tac0turtle) [reply store/rootmulti/snapshot_test.go:131]: if you run the test with go 1.18 it passes with the previous hash, but with go 1.19 it fails

**Comment by mattverse** [comment]
Seeing approval and same changes having gone through sufficient code review in upstream, gonna go ahead and merge this

**Comment by alexanderbez** [comment]
There are some failing tests, but yes once CI passes. We can merge 👍
